File: src/line_tokens.rb
module TokenBase
  def as_single_line
    self
  end

  def as_multi_line
    self
  end

  def is_a_comma?
    false
  end

  def is_a_newline?
    false
  end

  def is_a_comment?
    false
  end

  def is_keyword?
    false
  end

  def is_indent?
    false
  end

  def declares_class_or_module?
    false
  end

  def declares_if_or_unless?
    false
  end

  def is_end?
    false
  end

  def is_do?
    false
  end

  def is_def?
    false
  end

  def is_else?
    false
  end

  def is_require?
    false
  end

  def is_requirish?
    false
  end

  def is_private?
    false
  end

  def is_empty_string?
    false
  end
end

class HardNewLine
  include TokenBase
  def to_s
    "\n"
  end

  def is_a_newline?
    true
  end
end

class SoftNewLine
  include TokenBase
  def to_s
    "\n"
  end

  def as_single_line
    Space.new
  end

  def is_a_newline?
    true
  end
end

class CollapsingNewLine
  include TokenBase
  def to_s
    "\n"
  end

  def as_single_line
    DirectPart.new("")
  end

  def is_a_newline?
    true
  end
end

class DirectPart
  include TokenBase
  def initialize(part)
    @part = part
  end

  def to_s
    @part
  end

  def is_a_newline?
    @part == "\n"
  end

  def is_require?
    @part == "require"
  end

  def is_requirish?
    require_regex = /([^A-Za-z0-9]|^)require([^A-Za-z0-9])?/
    require_regex === @part
  end

  def is_private?
    @part == "private"
  end

  def is_empty_string?
    @part == ""
  end
end

class SingleSlash
  include TokenBase
  def to_s
    "\\"
  end
end

class Binary
  include TokenBase
  def initialize(symbol)
    @symbol = symbol
  end

  def to_s
    " #{@symbol} "
  end
end

class Space
  include TokenBase
  def to_s
    " "
  end
end

class Dot
  include TokenBase
  def to_s
    "."
  end
end

class LonelyOperator
  include TokenBase
  def to_s
    "&."
  end
end

class OpenParen
  include TokenBase
  def to_s
    "("
  end
end

class CloseParen
  include TokenBase
  def to_s
    ")"
  end
end

class OpenArgPipe
  include TokenBase
  def to_s
    "|"
  end
end

class CloseArgPipe
  include TokenBase
  def to_s
    "|"
  end
end

class DoubleQuote
  include TokenBase
  def to_s
    "\""
  end
end

class OpenSquareBracket
  include TokenBase
  def to_s
    "["
  end
end

class CloseSquareBracket
  include TokenBase
  def to_s
    "]"
  end
end

class Keyword
  include TokenBase
  def initialize(keyword)
    @keyword = keyword
  end

  def is_keyword?
    true
  end

  def declares_class_or_module?
    @keyword == :class || @keyword == :module
  end

  def declares_if_or_unless?
    @keyword == :if || @keyword == :unless
  end

  def is_end?
    @keyword == :end
  end

  def is_do?
    @keyword == :do
  end

  def is_def?
    @keyword == :def
  end

  def is_else?
    @keyword == :else
  end

  def to_s
    @keyword.to_s
  end
end

class Indent
  include TokenBase
  def initialize(spaces)
    @spaces = spaces
  end

  def to_s
    " " * @spaces
  end

  def is_indent?
    true
  end
end

class SoftIndent
  include TokenBase
  def initialize(spaces)
    @spaces = spaces
  end

  def to_s
    " " * @spaces
  end

  def as_single_line
    DirectPart.new("")
  end

  def is_indent?
    true
  end
end

class CommaSpace
  include TokenBase
  def to_s
    ", "
  end
end

class Comma
  include TokenBase
  def is_a_comma?
    true
  end

  def to_s
    ","
  end
end

class Op
  include TokenBase
  def initialize(op)
    @op = op
  end

  def to_s
    @op
  end
end

class Comment
  include TokenBase
  def initialize(content)
    @content = content
  end

  def is_a_comment?
    true
  end

  def to_s
    @content.to_s
  end
end

class CommentBlock
  include TokenBase

  attr_reader :comments

  def initialize
    @comments = []
  end

  def empty?
    @comments.empty?
  end

  def add_comment(comment)
    raise ArgumentError, "got something other than a comment" unless Comment === comment
    @comments << comment
  end

  def merge(other)
    @comments += other.comments
  end

  def to_token_collection
    TokenCollection.new(@comments.flat_map { |x| [x, HardNewLine.new] })
  end

  def is_a_comment?
    true
  end
end

File: native/src/line_tokens.rs
use crate::types::ColNumber;

#[derive(Debug, Clone)]
pub struct BreakableEntry {
    spaces: ColNumber,
    tokens: Vec<LineToken>,
}

impl BreakableEntry {
    pub fn new(spaces: ColNumber) -> Self {
        BreakableEntry {
            spaces,
            tokens: vec![],
        }
    }

    pub fn push(&mut self, lt: LineToken) {
        self.tokens.push(lt);
    }

    pub fn as_tokens(self) -> Vec<LineToken> {
        self.tokens
    }

    pub fn single_line_string_length(&self) -> usize {
        self.tokens.iter().map(|tok| {
            if tok.is_indent() {
                tok.clone().as_multi_line()
            } else {
                tok.clone().as_single_line()
            }
        }).map(|tok| tok.to_string().len()).sum()
    }
}

#[derive(Debug, Clone)]
pub enum LineToken {
    // this is all bodil's fault
    CollapsingNewLine,
    HardNewLine,
    SoftNewline,
    Indent { depth: u32 },
    SoftIndent { depth: u32 },
    Keyword { keyword: String },
    DirectPart { part: String },
    CommaSpace,
    Comma,
    Space,
    Dot,
    ColonColon,
    LonelyOperator,
    OpenSquareBracket,
    CloseSquareBracket,
    OpenParen,
    CloseParen,
    BreakableEntry(BreakableEntry),
    Op { op: String },
    DoubleQuote,
    LTStringContent { content: String },
    SingleSlash,
    Comment { contents: String },
}

impl LineToken {
    pub fn as_single_line(self) -> LineToken {
        match self {
            Self::CollapsingNewLine => LineToken::DirectPart {
                part: "".to_string(),
            },
            Self::SoftNewline => LineToken::Space,
            Self::SoftIndent { depth: _ } => LineToken::DirectPart {
                part: "".to_string(),
            },
            x => x,
        }
    }

    pub fn as_multi_line(self) -> LineToken {
        self
    }

    pub fn is_indent(&self) -> bool {
        match self {
            Self::SoftIndent{..} => true,
            Self::Indent{..} => true,
            _ => false,
        }
    }

    pub fn is_newline(&self) -> bool {
        match self {
            Self::HardNewLine => true,
            Self::SoftNewline => true,
            Self::CollapsingNewLine => true,
            Self::DirectPart { part } => {
                if part == "\n" {
                    panic!("shouldn't ever have a single newline direct part");
                } else {
                    false
                }
            }
            _ => false,
        }
    }

    pub fn to_string(self) -> String {
        match self {
            Self::CollapsingNewLine => "\n".to_string(),
            Self::HardNewLine => "\n".to_string(),
            Self::SoftNewline => "\n".to_string(),
            Self::Indent { depth } => (0..depth).map(|_| ' ').collect(),
            Self::SoftIndent { depth } => (0..depth).map(|_| ' ').collect(),
            Self::Keyword { keyword } => keyword,
            Self::DirectPart { part } => part,
            Self::CommaSpace => ", ".to_string(),
            Self::Comma => ",".to_string(),
            Self::Space => " ".to_string(),
            Self::Dot => ".".to_string(),
            Self::ColonColon => "::".to_string(),
            Self::LonelyOperator => "&.".to_string(),
            Self::OpenSquareBracket => "[".to_string(),
            Self::CloseSquareBracket => "]".to_string(),
            Self::OpenParen => "(".to_string(),
            Self::CloseParen => ")".to_string(),
            Self::BreakableEntry(BreakableEntry { spaces: _, tokens }) => {
                tokens.into_iter().fold("".to_string(), |accum, tok| {
                    format!("{}{}", accum, tok.to_string()).to_string()
                })
            }
            Self::Op { op } => op,
            Self::DoubleQuote => "\"".to_string(),
            Self::LTStringContent { content } => content,
            Self::SingleSlash => "\\".to_string(),
            Self::Comment { contents } => format!("{}\n", contents),
        }
    }

    pub fn is_single_line_breakable_garbage(&self) -> bool {
        match self {
            Self::Comma => true,
            Self::Space => true,
            Self::SoftNewline => true,
            Self::DirectPart{part} => (part == &"".to_string()),
            _ => false,
        }
    }
}
